client/src/client_read.c

```diff
@@ -597,9 +621,74 @@ int process_gfid_reads(unifyfs_client* client,
+        int num_request_selected = 0;
+        for (int i = 0; i < in_count; ++i) {
+            int fid = unifyfs_fid_from_gfid(client, in_reqs[i].gfid);
+            /* get meta for this file id */
+            unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(client, fid);
+            if (meta != NULL) {
+                if (!meta->attrs.is_laminated || !meta->needs_reads_sync) {
+                    /* do not proceed for this request as
+                     * it is not a laminated file or has already been synced.*/
+                    continue;
+                }
+                num_request_selected++;
+                off_t filesize_offt = unifyfs_gfid_filesize(client,
+                                                            in_reqs[i].gfid);
+                cur->value.file_pos = 0;
+                cur->value.length = filesize_offt - 1;
+                cur->value.gfid = in_reqs[i].gfid;
+                if (i < in_count - 1) {
+                    cur->next = calloc(1, sizeof(struct extents_list));
+                    cur->next->next = NULL;
+                    cur = cur->next;
+                } else {
+                    cur->next = NULL;
+                }
+                meta->needs_reads_sync = 0;
+            }
+        }
+        if (num_request_selected > 0) {
+            /* There are files which are laminated and
+             * require sync of extents */
+            size_t extent_count = 0;
+            unifyfs_client_index_t* extents = NULL;
+            int rc =
+                  invoke_client_node_local_extents_get_rpc(client,
+                                                           num_request_selected,
+                                                           list,
+                                                           &extent_count,
+                                                           &extents);
+            if (rc == UNIFYFS_SUCCESS && extent_count != 0) {
+                for (int j = 0; j < extent_count; ++j) {
+                    if (extents[j].log_app_id ==
+                        client->state.app_id) {
+                        int fid = unifyfs_fid_from_gfid(client,
+                                                        extents[j].gfid);
+                        /* get meta for this file id */
+                        unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(
+                                client,
+                                fid);
+                        if (meta != NULL) {
+                            seg_tree_add(&meta->extents,
+                                         extents[j].file_pos,
+                                         extents[j].file_pos +
+                                         extents[j].length - 1,
+                                         extents[j].log_pos,
+                                         extents[j].log_client_id);
+                        }
+                    }
+                }
+            }
+            free(extents);
```

Comment:
should probably move this inside the if block, or check for NULL before calling free()